### ISSUE TITLE
Biased avg loss reporting for mixed datasets

### ISSUE THREAD

**Issue by trufanov-nok** [JohnLangford/vowpal_wabbit#1213]
Hi,

By mixed dataset I mean dataset that contains both labeled and unlabeled examples. Usually one doesn't work with such dataset and has train and validation sets of labeled data and test set with unlabeled data. I'm applying VW's online learning to data generated at real time. A good example is a financial data. It can flow endlessly and you could test an unlabeled example, use prediction and then realize a real label in N minutes after that as market moves on, so you can use same example with label for training. And as the model is shared it has sense to use one VW instance for such case instead of 2 instances with shared model (one for training and one for prediction). So you will work with data flow with mixed labeled and unlabaled examples. And in described case it will be sequence of about 50% of labeled and 50% of unlabeled data mixed with a time lag.

Ok, I did following - saved a piece of such data for analysis and shuffled it just in case. Then played with it in VW and realized that the "best constant's loss" reported in VW output's footer is incorrect. And as this was mine feature commited years ago I've decided to fix it.
Then I did the same with mixed dataset but remove all features from it. So each example had only constant feature. That let me directly compare best constant loss and average loss as they should have pretty same value. And I noticed that my best constant loss is approx twice bigger.

After debugging I've concluded that the best constant's loss is fine and it's average loss that's biased here. Avg loss is calculated as roughly `sum_loss / weighted_examples` or `loss_sum / weighted_holdout_examples` for pass>1 and no holdout_off. And both `weighted_examples` and `weighted_holdout_examples` are calculated in `output_and_account_example() {all.sd->update()}` for *all* examples that have weight. And test examples have weight . So, while `loss_sum` contains only data for examples that have *both* weight and label, the `weighted_*` variables contains data for both labeled with weight and unlabeled data with weight.

That means if you have `N %` of unlabeled data in your dataset your real avg loss will be scaled by `N/100`. So in my case with ~50% of unlabeled data it's twice less than real one. 
This effect makes average loss be dataset dependent for mixed datasets and might spoil model performance comparison and hyperparameter search. 

It should be easy to fix. Instead of `sum_loss / weighted_examples` VW should use  `if (weighted_examples - weighted_unlabeled_examples + all.initial_t != 0) return sum_loss / (weighted_examples - weighted_unlabeled_examples + all.initial_t) else return 0.`. And VW's shared_data already has required ` weighted_unlabeled_examples` that's calculated as `all.sd->weighted_unlabeled_examples += ld.label == FLT_MAX ? ec.weight : 0;`.  But as for holdout case and `sum_loss / weighted_holdout_examples` it might require new variable for ` weighted_unlabeled_holdout_examples` equivalent.

**Comment by JohnLangford** [comment]
I fixed this one (finally!).  It was far from an easy fix, because I ended up refactoring the loss accounting which created cascading changes all over.  Nevertheless, the code is in a better state now.

**Comment by trufanov-nok** [comment]
Thank you!
